### PR TITLE
[Compose] Upgrade versions

### PR DESCRIPTION
**Pull request by ayush547**
### Problem 
Our compose/kotlin and other libraries used are out of date.

### Root cause 
Upgrades need to be tested and implemented.

### Fix
Checkout upto date version and their compatibilities and implement.
Impose compose bom to handle version codes on its own.

Other improvements -

- Replace hard coded versions that got missed with their respective consts
- Reorder version code consts in alphabetical order for easier tracking and updating
- Remove duplicate version codes consts - espressoVersion/roboelectricVersion
- Added waiters for open/close drawers for consistent tests
- Move core-kt from hard coded to const

### Validations
App runs fine
Tests run successfully.

### Screenshots

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Screenshot or description before this change | Screenshot or description with this change |

### Pull request checklist

This PR has considered:
- [ ] Light and Dark appearances
- [x] Automated Tests
- [ ] Documentation and demo app examples
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and RTL layouts
- [ ] Size classes and window sizes (notched devices, multitasking, different window sizes, etc)

### CODE REVIEW

File: fluentui_others/build.gradle [hunk 1]
```diff
@@ -27,6 +27,9 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
+    composeOptions {
```
Comment (mishramayank1) [inline fluentui_others/build.gradle:30]: this is not required as this module does not contain compose dependencies.

File: FluentUI.Demo/build.gradle [hunk 2]
```diff
@@ -85,19 +84,21 @@ dependencies {
...
+    implementation ("androidx.compose.material:material")
     implementation ("androidx.activity:activity-compose:$composeActivityVersion")
-    implementation 'androidx.test.ext:junit-ktx:1.1.3'
-    androidTestImplementation("androidx.compose.ui:ui-test-junit4:$composeVersion")
-    debugImplementation("androidx.compose.ui:ui-test-manifest:$composeVersion")
-    implementation "androidx.compose.runtime:runtime-livedata:$composeVersion"
+    implementation 'androidx.test.ext:junit-ktx:1.1.5'
+    androidTestImplementation("androidx.compose.ui:ui-test-junit4:$composeTestVersion")
+    debugImplementation("androidx.compose.ui:ui-test-manifest")
+    debugImplementation "androidx.test:monitor:1.6.1"
```
Comment (praveen970) [inline FluentUI.Demo/build.gradle:100]: is the version supposed to be hardcoded here?

File: FluentUI.Demo/build.gradle [hunk 3]
```diff
@@ -115,6 +116,7 @@ dependencies {
     androidTestImplementation "androidx.test.espresso:espresso-intents:${espressoVersion}"
     androidTestImplementation "androidx.test:runner:${androidxRunner}"
     androidTestImplementation "androidx.test:rules:${androidxRunner}"
+    androidTestImplementation "org.hamcrest:hamcrest:2.2"
```
Comment (praveen970) [inline FluentUI.Demo/build.gradle:119]: is the version supposed to be hardcoded here?

File: build.gradle [hunk 4]
```diff
@@ -30,35 +33,35 @@ allprojects {
...
-        roboelectricVersion       =    '4.4'
-        constraintLayoutVersion   =    '2.0.4'
+        androidxRunner            =    '1.5.0'
+        androidxCoreKt            =    '1.10.1'
+        appCompatVersion          =    '1.6.1'
+        buildToolVersion          =    '31.0.0'
+        composeActivityVersion    =    '1.7.2'
+        composeBomVersion         =    '2023.06.01'
+        composeTestVersion        =    '1.4.3'
+        composeVersion            =    '1.4.7'
```
Comment (praveen970) [inline build.gradle:47]: Do we need this when bom is used?

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/persistentbottomsheet/SheetHorizontalItemList.kt [hunk 5]
```diff
@@ -183,7 +183,10 @@ open class SheetHorizontalItemList @JvmOverloads constructor(context: Context, a
      */
     private fun setCollectionAccessibility(view: ViewGroup, size: Int) {
         ViewCompat.setAccessibilityDelegate(view, object : AccessibilityDelegateCompat() {
-            override fun onInitializeAccessibilityNodeInfo(host: View?, info: AccessibilityNodeInfoCompat?) {
+            override fun onInitializeAccessibilityNodeInfo(
+                host: View,
```
Comment (praveen970) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/persistentbottomsheet/SheetHorizontalItemList.kt:187]: Let remove this file from PR if there is no change

File: fluentui_icons/build.gradle [hunk 6]
```diff
@@ -47,11 +46,13 @@ android {
 
 dependencies {
     implementation fileTree(include: ['*.jar'], dir: 'libs')
-    implementation "androidx.compose.ui:ui:$composeVersion"
+    //Compose BOM
+    implementation platform("androidx.compose:compose-bom:$composeBomVersion")
```
Comment (praveen970) [inline fluentui_icons/build.gradle:50]: Do we need this here?

File: fluentui_peoplepicker/src/main/java/com/microsoft/fluentui/peoplepicker/PeoplePickerTextView.kt [hunk 7]
```diff
@@ -899,7 +899,7 @@ internal class PeoplePickerTextView : TokenCompleteTextView<IPersona> {
             setInfoText(info)
         }
 
-        override fun onPopulateAccessibilityEvent(host: View?, event: AccessibilityEvent?) {
+        override fun onPopulateAccessibilityEvent(host: View, event: AccessibilityEvent) {
```
Comment (praveen970) [inline fluentui_peoplepicker/src/main/java/com/microsoft/fluentui/peoplepicker/PeoplePickerTextView.kt:902]: Should we update v1 file?

File: FluentUI.Demo/build.gradle (hunk 2 above)
Comment (ayush547) [reply FluentUI.Demo/build.gradle:100]: Had to add this to fix some missing modules during test runs, these are present in this location only, and not intended to be used in other modules.
Do you recommend moving them out?

File: FluentUI.Demo/build.gradle (hunk 3 above)
Comment (ayush547) [reply FluentUI.Demo/build.gradle:119]: Same as above, added this to fix some missing modules during test runs, these are present in this location only, and not intended to be used in other modules.
Do you recommend moving them out?

File: build.gradle [hunk 8]
```diff
@@ -5,19 +5,22 @@
...
-    ext.kotlin_version = '1.6.20'
+    ext.kotlin_version = '1.8.21'
     ext.nexus_plugin_version = '0.30.0'
     ext.hydra_client_plugin = '1.1.4'
     
     repositories {
         google()
         jcenter()
         maven {
+            url "https://maven.google.com"
```
[the remark this answers is not in the document]
Comment (ayush547) [reply build.gradle:16]: No, checked that already

File: build.gradle (hunk 4 above)
Comment (ayush547) [reply build.gradle:47]: composeVersion is used for compiler version
composeTestVersion is used for compose test library version, it isn't working with bom

File: fluentui_icons/build.gradle (hunk 6 above)
Comment (ayush547) [reply fluentui_icons/build.gradle:50]: Yes, needed for the - androidx.compose.ui:ui dependency

File: build.gradle (hunk 4 above)
Comment (mishramayank1) [reply build.gradle:47]: Rename composeVersion  to composeCompilerVersion

File: fluentui_icons/build.gradle (hunk 6 above)
Comment (praveen970) [reply fluentui_icons/build.gradle:50]: This wasnt used in other modules

File: FluentUI.Demo/build.gradle (hunk 3 above)
Comment (praveen970) [reply FluentUI.Demo/build.gradle:119]: lets move the version out

File: FluentUI.Demo/build.gradle (hunk 2 above)
Comment (praveen970) [reply FluentUI.Demo/build.gradle:100]: lets move the version out

File: fluentui_icons/build.gradle (hunk 6 above)
Comment (ayush547) [reply fluentui_icons/build.gradle:50]: Other modules take a dependency on fluentui_core and get bom from there. fluentui_icon does not have this dependency (does not require) so bom had to be added separately.

**Review by praveen970 — APPROVED** [review]

**Pull request merged**